src/main/java/com/redhat/devtools/lsp4ij/features/formatting/LSPClientSideOnTypeFormattingTypedHandler.java

```diff
@@ -0,0 +1,277 @@
+        TextRange formatTextRange = null;
+
+        int offset = editor.getCaretModel().getOffset();
+        int beforeOffset = offset - 1;
+
+        // If appropriate, find the statement that was just terminated
+        FormattingScope formattingScope = formattingFeature.getFormatOnStatementTerminatorScope(file);
+        if (formattingScope == FormattingScope.STATEMENT) {
+            List<TextRange> selectionTextRanges = LSPSelectionRangeSupport.getSelectionTextRanges(file, editor, beforeOffset);
+            if (!ContainerUtil.isEmpty(selectionTextRanges)) {
+                // Find the closest selection range that is extended to line start/end; that should be the statement
+                Document document = editor.getDocument();
+                CharSequence documentChars = document.getCharsSequence();
+                formatTextRange = ContainerUtil.find(
+                        selectionTextRanges,
+                        selectionTextRange -> {
+                            int startOffset = selectionTextRange.getStartOffset();
+                            int endOffset = selectionTextRange.getEndOffset();
+
+                            // Remove leading newlines from the range
+                            while ((startOffset < endOffset) && (documentChars.charAt(startOffset) == '\n'))
+                                startOffset++;
+                            // Remove trailing whitespace up until newlines (inclusive) from the range
+                            boolean foundNewline = false;
+                            while (endOffset > startOffset) {
+                                char previousEndChar = documentChars.charAt(endOffset - 1);
+                                if (!Character.isWhitespace(previousEndChar)) {
+                                    break;
+                                } else if (previousEndChar == '\n') {
+                                    endOffset--;
+                                    foundNewline = true;
+                                } else if (!foundNewline) {
+                                    endOffset--;
+                                } else {
+                                    break;
+                                }
+                            }
+
+                            // See if this is a selection of complete lines
+                            int startLineNumber = document.getLineNumber(startOffset);
+                            int startLineStartOffset = document.getLineStartOffset(startLineNumber);
+                            if (startLineStartOffset == startOffset) {
+                                int endLineNumber = document.getLineNumber(endOffset);
+                                int endLineEndOffset = document.getLineEndOffset(endLineNumber);
+                                if ((endLineEndOffset == endOffset) || (endLineEndOffset == (endOffset + 1))) {
+                                    // Make sure that it ends with the terminator that was just typed
```

Comment:
If we found a contiguous range of full lines that (ignoring trailing whitespace) end in the statement terminator character that was just typed, this is the range of the statement to be formatted.